pkg/agent/agent.go

```diff
@@ -262,11 +269,76 @@ func (s *NimbessAgent) updatePipelinesEgress(port *network.EgressPort, excludedK
 	return nil
 }
 
+// removeEgressPort removes an Egress Port from a Meta Pipeline by name
+func (s *NimbessAgent) removeEgressPort(name string, metaKey string) error {
+	log.Debugf("Removing Egress Port: %s", name)
+	if len(s.MetaPipelines[metaKey].EgressPorts) == 1 {
+		if s.MetaPipelines[metaKey].EgressPorts[0].GetName() == name {
+			log.Debugf("Found egress port for removal from Meta Pipeline: %s", name)
+			s.MetaPipelines[metaKey].EgressPorts = nil
+			return nil
+		}
+		log.Error("Failed to find Egress Port in MetaPipeline")
+		return fmt.Errorf("unable to find egress port %s to remove in metapipeline with key: %s", name, metaKey)
+	}
+	for idx, ePort := range s.MetaPipelines[metaKey].EgressPorts {
+		if ePort.GetName() == name {
+			log.Debugf("Found egress port for removal from Meta Pipeline: %s", name)
+			s.MetaPipelines[metaKey].EgressPorts = append(s.MetaPipelines[metaKey].EgressPorts[:idx],
+				s.MetaPipelines[metaKey].EgressPorts[idx+1])
+			return nil
+		}
+	}
+	log.Error("Failed to find Egress Port in MetaPipeline")
```

Comment:
do we typically log an error and also pass the error down the line? Or should we let the caller log it if they want. If we do keep both, why not make them the same message?

Reply:
I like the lower level function logging the error. The reason the message is different is because line 274 already tells you in the log that which port is being removed. So there's no reason to print that again in the log. However, I did miss putting the metapipeline key in there. I'll add that.